### ISSUE TITLE
[Proposal] New migrate primitives to resolve failed migrations

### ISSUE THREAD

**Issue by floelhoeffel** [prisma/prisma#10561]
## tl;dr

Migrations can sometimes fail and this usually leads to situations tricky to solve. Some solutions offer automatic rollbacks but we believe they give a wrong sense of security as rollbacks might or might not succeed.

As we still wanted to improve our support around failed migrations, we decided to propose some new tooling around that and this document explains what we have in mind.

Please read through the proposed solution below and help us understand if this would be helpful to you.

For feedback, simply leave a comment. Alternatively, [book 30 minutes with us](https://calendly.com/loelhoeffel/talk-to-prisma-migrations-team). 

## Example of a failed migration

Imagine the following scenario: You created a migration with `prisma migrate dev` which was saved in a `*.sql` file containing several SQL statements. These statements are:

```sql
ALTER TABLE `Users` DROP `birthDate`;
ALTER TABLE `Users` ADD CONSTRAINT  `unique_emails` UNIQUE (`email`);
ALTER TABLE `Users` ADD COLUMN `signup_date` Date;
```

Applying the migration succeeded in your dev environment and now it is time to release to production. Unfortunately, the operation fails, since there is non-unique data in your production database (e.g. two users with the same email address).

Migrate returned the error directly from the database which lets you know about the violation of the unique constraint. 

```bash
ERROR 1062 (23000): Duplicate entry 'signeduptwice@gmail.com' for key 'unique_emails'
```

You now need to recover manually from the partially executed migration ("partially" as the removal of `birthDate` from table `Users` was successfully executed).

## Automatic rollbacks

Other tools sometimes auto-generate rollbacks. The problem with rollbacks is that they would not be of help here as your migration partially applied already. You now might argue that transactions could have helped but unfortunately transactions are not supported across different databases for DDL statements (SQL statements altering the schema). Additionally, they also come with performance penalties especially on larger datasets.

## Proposal: Migrate Primitives

We could offer migrate primitives to resolve failed migrations. Fundamentally you have two options how to proceed, forward and backwards:

- fix forward: apply the failed part of the migration or
- fix backwards: abort the migration and go back to the state before trying to run the failed migration

The two new commands we are using to fix the scenario from above are:

### prisma db execute

```bash
# Applies a SQL script to the database without interacting 
# with the prisma migrations table.
prisma db execute --target myDB --sql script.sql
```

### prisma migrate diff

```bash
# Diffs two diffable targets (prisma schema, database schema or
# migrations history). Output transforms one schema into another
# and can be human-readable or an executable SQL script.
prisma migrate diff --from $PRODUCTION_DATABASE_URL --to prisma/schema.prisma --output fix-migration.sql
```

### Move Forward

The existence of non-unique data in your database is unintentional and you want to fix that. Once fixed, you would be able to go ahead with the migration as planned:

- You need to figure out which rows contain duplicate values
- You need to delete / alter them to be able to apply the unique constraint
- You then can continue applying the rest of the failed migration
    - You **migrate** **diff** *from*: prodDB *to*: prismaSchema  *output*: migrationScript
    - You **db execute** the resulting script against your prod database
    - You **db migrate resolve —applied** to mark the previously failed migration as succeeded in prod

Your local migration history now yields the same result as the state your prod db is in. 

### Move Backwards

You realize that non-unique data is valid and you cannot move forward with the migration with the unique constraint included:

- You need to create a migration that takes your prod db to the state of your datamodel before the last migration
- You get the previous datamodel from git, or you get it from the local migrations history
    - You **migrate diff** f*rom*: prod *to*: prev datamodel *output*: migration script
    - You **db execute** that migration against prod.
    - You **db migrate resolve —rolled_back** the failed migration marking it as rolled back.
    - You now delete the folder of the failed migration to prevent it from being run locally again.

Your local migration history now yields the same result as the state your prod db is in. You can now modify the datamodel again to create a migration that suits your new understanding of the feature you're working on (with non-unique mail addresses).

## Conclusion

Resolving failed migrations is hard as every situation is different. Often, there is no automatic fix available through rollbacks. We believe offering the described commands gives users the right level of control to address their specific needs. 

We can't wait to hear what you think.

For feedback, [book 30 minutes with us](https://calendly.com/loelhoeffel/talk-to-prisma-migrations-team) or leave a comment.

Thanks

Edit 1: Updated `diff` example

**Comment by florianmartens** [comment]
@floelhoeffel thank you for working on this. I'll try to mirror my use-cases against this without claiming to match your knowledge about the underlying workings of Prisma.

When thinking of rollbacks two use cases come to my mind that would be really useful for me:

**1. Resolving failed or partially applied migrations**
The is described solution would meet my needs for resolving those situations well. If I understood the proposal correctly, this is likely to add a `scripts` folder to the local `prisma` directory. These scripts don't have an order to apply them in and are only useful when the target database is in certain states. To me, this would mean that these scrips should be short-lived and shouldn't necessarily be checked into version control (or live inside the project at all).

**2. Working on different features**
When working on different branches, I often have to apply a certain migration to develop a feature (e. g. add a column). When working on another branch, I need a way to quickly revert the database to the state of the new branch. Using the proposed solution, I would:
 - Diff the schema on the current branch against the state of the db
 - Apply the resulting migration script to the database
 This sounds great to me, but I am, again, left with an artifact (the migration script) that now lives in my project and is purely situational (it should probably be deleted after application).

A simple solution to this could be to have a simple flag that removes the script after it is applied (this is probably just a detail at this point).

**Two additional thoughts that came into my mind are:**
1. AFAIK the concept of a `target` database does not yet exist in Prisma. Currently, most Prisma applications connect to a database using a connecting string supplied by an environment variable. Does this proposal also include a way to dynamically select a target database based on environment variables?
2. Will diffing be able to resolve implicit and explicit m2m relations (implicit join tables)?

Again, thank you for your work on this. I'm really excited.

**Comment by maoosi** [comment]
This proposal looks great to address the issue I opened around SQL diffs https://github.com/prisma/prisma/issues/7738

I just have one question about the `prisma migrate diff` command. Do we imagine it possible to diff between two different formats, such as `prisma migrate diff —from mysql://... —to schema.prisma —output pretty|script.sql`?

Or would it be restricted to using the same diffable types for the `--from` and `--to` arguments? I think allowing the use of different types would make the proposal more flexible using the diff command outside of the failed migrations scenario described.

**Comment by floelhoeffel** [comment]
hey @florianmartens - please find my answers inline

> 1. Resolving failed or partially applied migrations
> The described solution would meet my needs for resolving those situations well. If I understood the proposal correctly, this is likely to add a scripts folder to the local prisma directory. This is the only caveat I see since these scripts don't have an order to apply them in and are only useful when the target database is in certain states. To me, this would mean that these scrips should be short-lived and shouldn't necessarily be checked into version control (or live inside the project at all).

We envision the output of `diff` to be short-lived. So you either pipe it into another command or you persist it locally. However we believe there is no good use case for keeping these scripts around. If you want a schema to be reflected in version control the answer is a migration and not the output of `diff`.

> 2. Working on different features
When working on different branches, I often have to apply a certain migration to develop a feature (e. g. add a column). When working on another branch, I need a way to quickly revert the database to the state of the new branch. Using the proposed solution.

Same as above: You would execute the output of diff against your database and move on. There should not be any impact on the migrations in your current branch.

> AFAIK the concept of a target database does not yet exist in Prisma. Currently, most Prisma applications connect to a database using a connection string supplied by an environment variable. Does this proposal also include a way to dynamically select a target database based on environment variables?

Great question - we could allow passing in a connection string directly into `diff` instead of just using whats in `schema.prisma`. We wouldnt try solve environments with this proposal. That is a separate feature.

**Comment by floelhoeffel** [comment]
hey @maoosi - we envision `diffable` to support:

- Prisma schema files
- Migrations history (e.g. files in your `migrations` folder)
  - maybe allowing to specify up to which migration e.g. `HEAD-1` or something similar
- any database we support

Diffing between databases would work as long as the connector is the same. E.g. you could diff a MariaDB and a MySQL database as they use the same prisma connector under the hood.

**Comment by m1212e** [comment]
Hi, I think this proposal is a very good idea. I'm not super deep into how prisma works under the hood.
But coming from #8056 just a quick question: Why the need for a running database to generate the actual migration schemas?
In our current setup the database lives in a private network with the REST endpoints we deployed and developers don't have direct access. Currently we all spin up a local DB when we made changes to the prisma schema and then generate the actual migrations. With a possibility to run e.g. prisma migrate dev --create-only **--force-schemas** to only create the schemas which then can be checked in to version control this process could be automated in a precommit hook or something similar.
I really appreciate all the work you put into this awesome tool, thanks a lot!

**Comment by floelhoeffel** [comment]
@m1212e - thanks for the input and two quick questions:
- what would be the effect of `--force-schemas`?
- how do you apply the migrations to your prod database if the devs dont have access to it? Would be awesome if you could describe that workflow end2end 🙇

**Comment by m1212e** [comment]
That was just a suggestion, you guys will probably find a better name :D But basically it will generate the actual migration files, without requiring a running database to connect to. Currently I need to start an empty mysql db locally only to generate the migration files. With the flag enabled, it would just assume that the db type specified should be used to create the sql files and output them into the migration directory.

Sure: So with our current setup we got our endpoints and database inside a network on our servers. The firewall blocks all access besides http(s). Whenever a  dev makes changes to the schema it runs through all tests etc. and if it isn't breaking anything and no test fails, the ci/cd server deploys and then sends an http request to a specific endpoint, which is starting the migration of the production database. That way we have direct feedback of potentially failing migrations inside our ci/cd.
The reason for this process is beyond my knowledge, actually I found it pretty cumbersome.
My plan for migrating to prisma initially was to upload the whole prisma cli to that http endpoint and then call migrate indirectly via http. A migration API would be helpful but is probably beyond the scope of this issue.

**Comment by floelhoeffel** [comment]
@m1212e 

> That was just a suggestion, you guys will probably find a better name :D But basically it will generate the actual migration files, without requiring a running database to connect to. Currently I need to start an empty mysql db locally only to generate the migration files. With the flag enabled, it would just assume that the db type specified should be used to create the sql files and output them into the migration directory.

This would be solved by this proposal!

> Sure: So with our current setup we got our endpoints and database inside a network on our servers. The firewall blocks all access besides http(s). Whenever a dev makes changes to the schema it runs through all tests etc. and if it isn't breaking anything and no test fails, the ci/cd server deploys (not sure how, probably ssh but devs are denied access) and then sends an http request to a specific endpoint, which is starting the migration of the production database. That way we have direct feedback of potentially failing migrations inside our ci/cd. The reason for this process is beyond my knowledge, actually I found it pretty cumbersome. My plan for migrating to prisma initially was to upload the whole prisma cli to that http endpoint and then call migrate indirectly via http. A migration API would be helpful but is probably beyond the scope of this issue.

Very interesting, thanks for sharing. If you happen to learn more about the reasoning and/or if you wanted to share more about whats cumbersome to you, I am more than happy to jump on a call or read more here. Thanks!

**Comment by do4gr** [comment]
We've spend some time fleshing out the design for the proposed commands. Below is what we plan to implement for the initial version. Please let us know if you have any more feedback. 


## prisma db execute

### Helptext

Preview Feature (Warning)

Run a database native database command on the specified database. You need to specify the datasource and the command input. 

The output of the command is connector-specific, and is not meant for returning data, but only to report success or failure.

On SQL databases, this command takes as input a SQL script. The whole script will be sent as a single command to the database.

This command is currently not supported on MongoDB.

Examples:

```bash
# Standard input as source and explicit database url.
$ echo 'SELECT `hello world`;' | \
  prisma db execute \
  --preview-feature \
  --stdin \
  --url="mysql://home/testdb"

# SQL file as input and database url from schema.
script.sql
 'SELECT `hello world`;'

$ prisma db execute \
  --preview-feature \
  --file-path ./script.sql \
  --schema schema.prisma
```

### Arguments

- The datasource:
    - `--url`: the URL of the datasource to run the command on.
    - `--schema`: path to the Prisma schema file to take the datasource URL from.
    - Exactly one of `--url` or `--schema` must be provided. None is an error, and more than one is also an error.
- The command input:
    - `--file-path` : the path towards the file containing the commands
    - `--stdin` : take input from stdin
    - Exactly one of `--file-path` or `--stdin` must be provided. None is an error, and more than one is also an error.

### Output

- Free form, depending on the connector.
- Exit code:
    - 0: ok
        - 26 statements executed in 1.4s? (MySQL only)
    - 1: error (red)

### Error Handling

- engine will error on mongoDB for example

## prisma migrate diff

### Helptext

Compares the database schema from two arbitrary sources, and displays the differences either as a human-readable summary (by default) or executable script (For SQL databases: SQL, not supported on MongoDB).

[Link to expanded use cases in docs]

`prisma migrate diff` is a read-only command that does not write anything to any database (except the shadow database in special cases, see below).

A `diff` command takes a source (`--from`) and a destination (`--to`) and compares the source with the destination to generate the diff. The diff can be interpreted as generating a migration that brings the source schema (from) to the shape of the destination schema (to). The following targets are supported:

- `--from-url` / `--to-url` : a connection string pointing to a live database.
- `--from-empty` /`--to-empty` : an empty schema (takes no parameter)
- `--from-datamodel`/ `--to-datamodel`: the datamodel in a schema file (takes a path).
- `--from-schema`/`--to-schema`: the database specified via the datasource in a schema file (takes a path).
- `--from-migrations`/ `--to-migrations`: a migrations directory path used with Prisma migrate.

`prisma migrate diff` will ask you to provide a `--shadow-database-url` argument when necessary. For example when you diff migration directories with `--from-migrations` or `--to-migrations`, Prisma needs a **shadow database** to determine a schema from the migration files. You can read more about shadow databases at [link]

Examples:

```bash
Starting point options:
--from-empty
--from-schema prisma/schema.prisma
--from-datamodel prisma/schema.prisma
--from-url postgres://....
--from-migrations ./migrations...

End point options:
--to-empty
--to-schema prisma/schema.prisma
--to-datamodel prisma/schema.prisma
--to-url postgres://....
--to-migrations ./migrations...

Shadow db:
optional in most cases
--shadow-database-url "$SHADOW_DB"

Output format:
defaults to human readable summary on stdout
--script will render a sql script to stdout

Examples

# From database to database as summary
# e.g. compare two live databases
$ prisma migrate diff \
  --preview-feature \
  --from-url "$DATABASE_URL" \
  --to-url "postgresql://login:password@localhost:5432/db2"

# From empty to current datamodel as a SQL script
# e.g. to squash migrations
$ prisma migrate diff \
  --preview-feature \
  --from-empty \
  --to-datamodel schema.prisma \
  --script

# From current database specified in the schema to empty as a SQL script
# e.g. to teardown a database
$ prisma migrate diff \
  --preview-feature \
  --from-schema schema.prisma \
  --to-empty \
  --script

# From a live db to a Prisma datamodel
# e.g. roll forward after a migration failed in the middle
$ prisma migrate diff \
  --preview-feature \
  --shadow-database-url "$SHADOW_DB" \
  --from-url "$PROD_DB" \
  --to-datamodel=next_datamodel.prisma \
  --script

# From a live database to a datamodel
# e.g. roll backward after a migration failed in the middle
$ prisma migrate diff \
  --preview-feature \
  --shadow-database-url "$SHADOW_DB" \
  --from-url "$PROD_DB" \
  --to-datamodel=previous_datamodel.prisma \
  --script

# From migrations directory to another database 
# e.g. generate a migration for a hotfix already applied on production
$ prisma migrate diff \
  --preview-feature \
  --shadow-database-url "$SHADOW_DB" \
  --from-migrations ./migrations \
  --to-url "$PROD_DB" \
  --script
```

### Arguments

- Diff targets:
    - `--from` / `--to`
        - `schema` — url from schema, takes an optional file path argument, but defaults to the project’s schema.prisma
        - `datamodel` — datamodel from schema, takes an optional file path argument, but defaults to the project’s schema.prisma
        - `empty` — an empty database
        - `url` — the schema of the database behind the url
        - `migrations` — can also target up to a specific migration (later)
- `--shadow-database-url`:
    - required when you use `--from-migrations` or `--to-migrations`
    - or are on a cloud provider for example
- ``--script``
    - defaults to best-effort human-readable summary of the diff
    - if `--script` is passed - a script executable on the datasource to bring from `--from` to `--to`

### Output

- The summary or script goes to stdout. This is the only output in the success case.
- Exit code:
    - 0: ok
    - 1: error (red)

**Comment by Jolg42** [comment]
This proposal is now closed as we just shipped `migrate diff` and `db execute` in Preview with `3.9.0`.
[See the releases notes here](https://github.com/prisma/prisma/releases/tag/3.9.0)

We are looking forward to learning how they work for you and how we could make them better.
[You are welcome to post feedback here](https://github.com/prisma/prisma/issues/11514)

And feel free to open an issue if you find a bug 🐛
